gravitee-io/gravitee-access-management
# feat(management): Identifier-first login

fixes gravitee-io/issues#5388

How to test: 
- Create an Access Management domain
- Create an application (make sure that you have the openid scope at least)
- Activate Identifier-first login at application (or settings) level

![localhost_4200_environments_default_domains_social-domains_settings_login (1)](https://user-images.githubusercontent.com/8531515/133396063-fc33b07a-8a5a-4496-a616-0a0e2e07f256.png)

- Create an identity provider with a or several `Domain Whitelist` patterns corresponding to the domain of the external identity provider, this will enable email domain lookup on the username (for example here, if the email domain matches `gmail.com` or `graviteesource.com`)

![localhost_4200_environments_default_domains_social-domains_applications_fbbc76c2-90b2-47a0-bc76-c290b267a039_settings_login (1)](https://user-images.githubusercontent.com/8531515/133268063-86a51cf0-a07f-460d-88cc-a57a35194756.png)

- Initiate the Login Flow, if the username is just a login / is an email not matching the domain whitelist of the configured providers you'll get to the 2nd step with a password to enter and the possible external identity providers

![localhost_8092_social-domains_2step_login_client_id=8dfff187-79ec-4eb0-bff1-8779ec6eb06b response_type=code redirect_uri=https%3A%2F%2Fnowhere com (1)](https://user-images.githubusercontent.com/8531515/133270104-7e2985d7-6a85-4076-be2e-87447e667de5.png)

![localhost_8092_social-domains_login_username=remi sultan%40domain com response_type=code redirect_uri=https%3A%2F%2Fnowhere com client_id=8dfff187-79ec-4eb0-bff1-8779ec6eb06b](https://user-images.githubusercontent.com/8531515/133270505-9b4510e6-a475-4c18-9510-47afa28b0cf0.png)

If the email matches one of the identity providers you'll get redirected to it
![image](https://user-images.githubusercontent.com/8531515/133270695-53145c84-b081-434b-9569-aaa1194951f2.png)

After this you should get the consent page or be logged directly if you are already logged in.

You can also edit the first screen login page at domain level or application level : 
![localhost_4200_environments_default_domains_social-domains_applications_fbbc76c2-90b2-47a0-bc76-c290b267a039_settings_general (1)](https://user-images.githubusercontent.com/8531515/133271306-e9ea480f-d81a-482e-855c-4b1fc0ba070d.png)

![localhost_4200_environments_default_domains_social-domains_applications_fbbc76c2-90b2-47a0-bc76-c290b267a039_settings_general](https://user-images.githubusercontent.com/8531515/133271113-6fbb7acd-8bd2-4d85-a853-e1bd61bd9a6a.png)

![localhost_8092_social-domains_2step_login_client_id=8dfff187-79ec-4eb0-bff1-8779ec6eb06b response_type=code redirect_uri=https%3A%2F%2Fnowhere com (2)](https://user-images.githubusercontent.com/8531515/133271467-528d41bd-70f0-4cee-a69d-2c087aaecc4c.png)


- Note: 
    - Hide Login is disabled since we need to redirect if the password is necessary
    - Works with WebAuthn  if the user already has already logged in with it.
    - User Registration can be activated as well


Tested with:
- MongoDB ✅ 
- Postges 13 ✅ 
- Postgres 9 ✅ 
- MySQL 8 ✅ 
- MySQL 5.7 ✅ 
- MariaDB 10.5 ✅  
- MariaDB 10.3 ✅ 
- MsSQL 2019 ✅ 
- MsSQL 2017 ✅
